Indexes 5: Adds `spk repo index` subcommand for index generation and updates (#1340)
spkenv/spk: docs/ref/indexes.md

---
title: Indexes
summary: Indexes for improving solve times
weight: 120
---

This explains indexes, indexing, index controls, and index requirements in `spk`.

## Spk Repository Indexes

`SPK` supports generating a packages index for a repository to help
with solves. An index must have been generated separately for a
repository before it can be used. Using an index speeds up solves

> against that use that repository.

Reviewer: ```suggestion
that use that repository.
```

Author: I've updated this.

The index is designed to help the solvers with solves (package
look-ups, deprecation checks, pre-loading non-package variable
references, getting install requirements, etc.). It does not contain
the full package data. So it does not have the information needed to
help other `spk` operations, e.g. building or testing a package.

> If indexing is enabled, you have to generate an index before to trying

Reviewer: ```suggestion
If indexing is enabled, you have to generate an index before trying
```

Author: I've updated this.

to use it in a solve. They are not generated on the fly (outside of
automated tests).

If index use is enabled, but no index has been generated for a
repository, `spk` will fallback to using the underlying repository's
packages directly. This typically results in slower solves,
particularly with larger repositories.

If a solve uses multiple repositories and indexes exist for some or
all of them, `spk` will use the indexes that exist.

### Enabling/Disabling Indexes

`spk` index use can be enabled or disabled in the `spk` config file.
It is disabled by default because index generation and updating needs

> to be set up on per repository basis. Setting this up is recommended

Reviewer: ```suggestion
to be set up on a per repository basis. Setting this up is recommended
```

Author: I've updated this.

for repositories with a large number of packages.

Most `spk` commands can use the `--use-indexes` and `--no-indexes`

> flags to override repository index use.`spk repo index ...` command

Reviewer: ```suggestion
flags to override repository index use. The `spk repo index ...` command
```

Author: I've updated this.

This will generate a flatbuffer schema based index file. The index
file is stored in the underlying spfs repo (e.g. origin repo) in a
`index/spk/` sub-directory.

Index generation only works on one repository at a time (hence the
`--disable-repo local` when working on the default `origin` repo). If
you have multiple repositories to index you have to run `spk repo index
...` once for each repository.

See `spk repo index -h` for more details.

You don't have to generate a full index every time a package changes,
you can also update a package in an existing index, see the next
section.

### Updating an existing Index

Updating a package in an existing index, such as after a new build is
published or a package is deprecated, is not automatic.

A site using `spk` has to set up a system to trigger index updates
when they want them to happen. The recommendation is after a new
package is published, deprecated, undeprecated, or deleted. But
regular periodic complete index generation may also work for a site,
depending on the frequency of package changes and the periodic updates.

To update an existing index, e.g. after a new `python` package was
published, run:

> `spk repo index --disable-repo local --update python`

Reviewer: A thought on usage, rather than being expected to remember to add `--disable-repo local` (because something bad would happen if not??), how about we make this `index` sub-command require specifying the repo to operate on, and error if one isn't specified.

`spk repo index -r origin [--update python]`

Author: The usage followed the existing `spk repo` command's subcommand, but I agree it is cumbersome.

- [x]  Update the usage of the `spk repo index` command to match this

Author: I've changed it to only take a `-r <reponame>` option and not all the other repo flags.

This will read in the existing index for the repository and update the
versions and builds of the named package in the index. It is faster
than generating an index from scratch. It has to be run once per
repository and once per package.

See `spk repo index -h` for more details.

## Index vs Repository mismatches - updates are important

When index use is enabled, it is important to update the indexes when
changes happen to the repository and its packages. Otherwise, the
index and real package data will get out of sync. An index that is old

> won't have the same data as the repositories it is based one. This can

Reviewer: ```suggestion
won't have the same data as the repositories it is based on. This can
```

Author: I've updated this.

lead to discrepancies in solves.

## Index Details

### Index file location

An index is stored in a file inside the repository it indexes. Only
spk filesystem repositories (based on spfs fs repositories) support
indexes. The file will be kept in the `index/spk/` sub-directory of an
spfs fs repository.

### Structure and types in SPK

An index is implemented by a set of types in spk-storage and

> spk-schema that work together to wrap a repository in its index and

Reviewer: ```suggestion
spk-schema that work together to wrap a repository with its index and
```

Author: I've updated this.

produce indexed packages that act as packages for the solver.
 
The hierarchy of index related types:

```
in the 'spk-storage' crate:

RepositoryHandle enum
   Indexed struct (implements Repository and Storage traits)
       RepoIndex enum (implements RepoIndex and RepoIndexMut traits)
           FlatBufferRepoIndex struct (implements RepoIndex and RepoIndexMut traits)
               flatbuffers schema

in the 'spk-schema' crate:
   IndexedPackage (implements Package traits, produced by a FlatBufferRepoIndex)
       flatbuffers schema
       
```

### Flatbuffer index schema

`spk` uses flatbuffers for the index data format on disk. This is fast
to read and use, but slow to generate. Updates to an existing index
require an index to be read in completely and written out again with
the updated data. It does not support updating complex structures
in-place.

The `spk-proto` crate contains spk's flatbuffers schema for an
index. It requires `flatc` to be installed to generate the rust code
for the index schema.

> The schema is versioned internally, and in the index file name contains

Reviewer: ```suggestion
The schema is versioned internally, and the index file name contains
```

Author: I've updated this.

the schema version number as well.

The index structure does not match the rust object structures

> 1-to-1. The is partially due to only keeping things needs for solves,

Reviewer: ```suggestion
1-to-1. This is partially due to only keeping things needed for solves,
```

Author: I've updated this.

                                      build digest/id
                                      ... other fields needed for solving
   list of GlobalVar  (used to prime the 'resolvo' solver to avoid restarts)
               name
               list of valid values
```

The structure is quite deep to cover the data for options,
requirements, compat rules, version numbers, build ids, and
components. See the schema itself for more detailed information,
including which fields of rust objects are being ignored, omitted from
an index.

### How to evolve the index schema

The flatbuffers data format supports adding new fields and structures
without breaking existing code, provided fields are not deleted. New
fields should be added only to the ends of tables.

The index schema's version number lets `spk` check that the index data
is compatible with the current spk being run. That is, that the code
understands this version of the schema. There will be situations where
multiple versions of spk are being used in a site and they may not all
be able to use index data generated by other versions of spk.

Spk includes the schema version number in the index file name to allow

> multiple index schema to co-exist during a transition between index

Reviewer: ```suggestion
multiple index schemas to co-exist during a transition between index
```

Author: I've updated this.

versions.

#### Adding a new field

When a new field needs to be added to the index schema, the developer should:
- Increment the schema version number
- Add the new field to the schema
- Increment spk's compatible schema version number
- Add spk code to read and write the new field
- Make a new release of spk
- Generate a new index using the new spk build
- Transition the site to the new spk release
- Once the transition is complete, retire the older index data file and generation

During the transition the older version of spk will use the older
index format(s) and the newer spk will use the newer one. Both indexes
will have to be generated and kept up to date during the transition.
Once the older spk version is fully retired, the older index data can
be retired too.

#### Stop using an existing field

When an older field should not be used anymore, the developer should:
- Add spk code to stop reading the old field
- Make a new release of spk that no longer reads the field
- Transition the site to the new spk release

There's no need to change the index schema version for to stop reading
a field, and index generation remains unchanged (even though it

> includes data that isn't read by spk anymore),

Reviewer: ```suggestion
includes data that isn't read by spk anymore).
```

Author: I've updated this.

> #### Removing an old field

Reviewer: I'm not sure the section above is useful, but removing a field (or stopping to populate an existing field), are reasons to bump the version number.

Author: I've added another section on stopping to populate an existing field. I felt I needed to make it clearer that it isn't the same as removing a field and its implications for reading that field in older spks. I've left the stopping to read a field in as a result and for completeness. Let me know if it's too verbose or still too much.

large number of packages.

Most `spk` commands can use the `--index-use <value>` option to override
repository index use on a per command basis. The `spk repo index ...`
command always has index use disabled because it operates on the
indexes themselves. `spk info ...` also disables indexes because
displaying complete package information requires reading the package
from the repository directly, not from an index.

If index use is enabled in the config file, it can be disabled with
the `--index-use disabled` command line option.

### Generating an Index

To generate an index for a repository (e.g. origin), run:
`spk repo index -r origin`

This will generate a flatbuffer schema based index file. The index
file is stored in the underlying spfs repo (e.g. origin repo) in a
`index/spk/` sub-directory.

Index generation only works on one repository at a time. If you have
multiple repositories to index you have to run `spk repo index ...`
once for each repository.

See `spk repo index -h` for more details.

You do not have to generate a full index every time a package changes,
you can update a package in an existing index, see the next section.

> ### Updating an existing Index

Reviewer: ```suggestion
### Updating an Existing Index
```

Author: I've updated this.

Updating a package in an existing index, such as after a new build is
published or a package is deprecated, is not automatic.

A site using `spk` has to set up a system to trigger index updates
when they want them to happen. The recommendation is after a new
package is published, deprecated, undeprecated, or deleted. But
regular periodic complete index generation may also work for a site,
depending on the frequency of package changes and the periodic updates.

To update an existing index, e.g. the `origin` repo's index after a
new `python` package was published or deprecated, run:
`spk repo index -r origin --update python`

The `--update` option can be given multiple times to update several
packages at a once, e.g.:
`spk repo index -r origin --update python --update zlib`

> The `--update` option take a package/version as well. This lets the

Reviewer: ```suggestion
The `--update` option takes a package/version as well. This lets the
```

Author: I've updated this.

update be restricted to a specific version of a package. This can make

> for shorter update times for packages with large numbers of versions,

Reviewer: ```suggestion
for shorter update times for packages with a large number of versions,
```

Author: I've updated this.

or builds per version, e.g.: 
`spk repo index -r origin --update python/3.10.8 --update zlib/1.2.12`

Those commands will read in the existing index for the repository and
update the versions and builds of the named package in the index. It
is faster than generating an index from scratch. It has to be run once

> per repository to update the given package or packages in that

Reviewer: ```suggestion
per repository to update the given package or packages in each
```

Author: I've updated this.

repository's index.

See `spk repo index -h` for more details.

> ## Index vs Repository mismatches - updates are important

Reviewer: The heading capitalization in this document is inconsistent (or I'm not seeing the pattern). Please pick one style.

Author: I've updated this to be more consistent.

When index use is enabled, it is important to update the indexes when
changes happen to the repository and its packages. Otherwise, the
index and real package data will get out of sync. An index that is old
will not have the same data as the repositories it is based on. This
can lead to discrepancies in solves.

## Index Details

### Index file location

An index is stored in a file inside the repository it indexes. Only
spk filesystem repositories (based on spfs fs repositories) support
indexes. The file will be kept in the `index/spk/` sub-directory of an
spfs fs repository.

> ### Structure and types in SPK

Reviewer: Another consistency nit, we can't decide if it's `Spk` or `SPK`.

Author: I've got with SPK when talking about SPK, and `spk` in the example command lines.
